### PR TITLE
Disk usage metrics and health indicators

### PR DESCRIPTION
**Pull request by stefanvozd**
Added disk usage metrics and health indicators

### CODE REVIEW

**Review by abuijze — CHANGES_REQUESTED** [review]
Minor change in logging requested

File: axonserver/src/main/java/io/axoniq/axonserver/config/AxonServerDiskMetricsConfig.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ */
+@Configuration
+@DependsOn("eventStoreManager")
+public class AxonServerDiskMetricsConfig {
+
+    private final Logger logger = LoggerFactory.getLogger(AxonServerDiskMetricsConfig.class);
+
+    private final LocalEventStore localEventStore;
+
+    public AxonServerDiskMetricsConfig(LocalEventStore localEventStore) {
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/config/AxonServerDiskMetricsConfig.java:29]: Disk space metrics may not be relevant for all types of event stores, The metrics are only relevant for the current implementation of the EventStorageEngine.

File: axonserver/src/main/java/io/axoniq/axonserver/config/DiskSpaceHealthIndicatorConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            this.threshold = threshold;
+        }
+
+        @Override
+        protected void doHealthCheck(Health.Builder builder) {
+            long diskFreeInBytes = this.path.getUsableSpace();
+            if (diskFreeInBytes >= this.threshold.toBytes()) {
+                builder.up();
+            } else {
+                logger.warn(String.format("Free disk space below threshold. Available: %d bytes (threshold: %s)", diskFreeInBytes, this.threshold));
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/config/DiskSpaceHealthIndicatorConfiguration.java:47]: Add this.path to logging to show for which path the disk space is running out.

File: axonserver/src/main/java/io/axoniq/axonserver/config/DiskSpaceHealthIndicatorConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                builder.status(new Status("WARN"));
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/config/DiskSpaceHealthIndicatorConfiguration.java:48]: Move class io.axoniq.axonserver.enterprise.HealthStatus from EE to SE, and use HealthStatus.WARN_STATUS.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/EventStorageEngine.java [hunk 4]
```diff
@@ -182,8 +184,9 @@ void processEventsPerAggregateHighestFirst(String aggregateId, long minSequenceN
     /**
      * Adds information to the actuator health endpoint for this event store.
      * @param builder actuator health builder
+     * @param diskSpaceHealthIndicatorProperties properties or health indicators
      */
-    default void health(Health.Builder builder) {
+    default void health(Health.Builder builder, DiskSpaceHealthIndicatorProperties diskSpaceHealthIndicatorProperties) {
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/EventStorageEngine.java:189]: These properties should not be part of the health operation, but set on the EventStorageImplementation level.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java [hunk 5]
```diff
@@ -469,26 +472,44 @@ private int retrieveEventsForAnAggregate(long segment, List<Integer> indexEntrie
...
+            }
+
+            builder.withDetail(context + ".total", store.getTotalSpace());
             builder.withDetail(context + ".free", store.getUsableSpace());
             builder.withDetail(context + ".path", path.toString());
+            builder.withDetail(context + ".threshold", threshold);
         } catch (IOException e) {
             logger.warn("Failed to retrieve filestore for {}", path, e);
         }
         while (n != null) {
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java:498]: this while loop can be removed, was part of an initial implementation of multi-location storage that is no longer implemented.

**Review by MGathier — CHANGES_REQUESTED** [review]
For the health indicator I would create a bean AxonServerDiskSpaceHealthIndicator extending 
the spring DiskSpaceHealthIndicator, and have the PrimaryEventStore instances register a location 
with this health indicator bean. This would make the whole implementation of the disk space health check a lot easier, and allow EE to register another FileSystem for the replication log. 

For the metrics I would let the PrimaryEventStore register the DiskSpaceMetrics.

File: axonserver/src/main/java/io/axoniq/axonserver/config/AxonServerDiskMetricsConfig.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+     * @param prometheusMeterRegistry micrometer Prometheus registry used for registering gRPC metrics
+     */
+    @Autowired
+    public void configureDiskMetrics(
+            PrometheusMeterRegistry prometheusMeterRegistry) {
+
+        localEventStore.registerFileStoreListener(fileStore -> {
+            try {
+                FileSystem fileSystem = fileStore.getFileSystem();
+                DiskSpaceMetrics diskSpaceMetrics = new DiskSpaceMetrics(fileSystem.getPath("/").toFile(), Tags.of("fileSystem", fileSystem.toString()));
```
[the remark this answers is not in the document]
Comment (stefanvozd) [reply axonserver/src/main/java/io/axoniq/axonserver/config/AxonServerDiskMetricsConfig.java:45]: should I add some other tag instead?

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/EventStreamReader.java [hunk 7]
```diff
@@ -10,6 +10,7 @@
 package io.axoniq.axonserver.localstorage;
 
 import io.axoniq.axonserver.grpc.event.EventWithToken;
+import org.springframework.boot.actuate.autoconfigure.system.DiskSpaceHealthIndicatorProperties;
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/EventStreamReader.java:13]: unused import

File: axonserver/src/test/java/io/axoniq/axonserver/localstorage/FakeEventStore.java [hunk 8]
```diff
@@ -13,6 +13,7 @@
 import io.axoniq.axonserver.grpc.event.EventWithToken;
 import org.springframework.data.util.CloseableIterator;
 
+import java.nio.file.Path;
```
Comment (MGathier) [inline axonserver/src/test/java/io/axoniq/axonserver/localstorage/FakeEventStore.java:16]: unused import

File: axonserver/src/test/java/io/axoniq/axonserver/rest/HttpStreamingQueryTest.java [hunk 9]
```diff
@@ -31,6 +31,7 @@
 import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;
 
 import java.io.IOException;
+import java.nio.file.Path;
```
Comment (MGathier) [inline axonserver/src/test/java/io/axoniq/axonserver/rest/HttpStreamingQueryTest.java:34]: unused

**Review by MGathier — APPROVED** [review]

**Pull request merged**